src/util/misc.rs

```diff
@@ -255,6 +306,77 @@ pub fn signed_msg_hash(msg: &str) -> sha256d::Hash {
     sha256d::Hash::from_engine(engine)
 }
 
+/// Ripemd160 hash of sha256 hash of given data
+pub fn hash160(data: &[u8]) -> ::hashes::hash160::Hash {
+    let mut sha_engine = sha256::Hash::engine();
+    sha_engine.input(data);
+
+    ::hashes::hash160::Hash::from_engine(sha_engine)
+}
+
+/// Convert a byte array of a pubkey hash into a segwit redeem hash
+pub fn segwit_redeem_hash(pubkey_hash: &[u8]) -> ::hashes::hash160::Hash {
+    let mut sha_engine = sha256::Hash::engine();
+    sha_engine.input(&[0, 20]);
+    sha_engine.input(pubkey_hash);
+    ::hashes::hash160::Hash::from_engine(sha_engine)
+}
+
+/// Pull out payload as byte array
+pub fn get_payload_bytes(address: &Address) -> &[u8] {
+    match &address.payload {
+        Payload::ScriptHash(hash) => &hash,
+        Payload::PubkeyHash(hash) => &hash,
+        Payload::WitnessProgram { program, .. } => &program,
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+/// Errors from failures to decode bech32
+pub enum Bech32DecodingError {
+    /// Excessive padding when converting Vec<u5> to Vec<u8>
+    ExcessivePadding,
+    /// Non-zero padding when converting Vec<u5> to Vec<u8>
+    NonZeroPadding,
+    /// The Bech32 is invalidly encoded
+    InvalidEncoding(bech32::Error),
+}
+
+/// decode address to Bech32 u8 byte array
+pub fn bech32_decode(address: &Address) -> Result<Vec<u8>, Bech32DecodingError> {
+    match bech32::decode(&address.to_string()) {
+        Ok((_, u5_vec, _)) => bech32_from_words(&u5_vec[1..]),
+        Err(e) => Err(Bech32DecodingError::InvalidEncoding(e))
+    }
+}
+
+/// Convert u5 vec into u8 vec
+fn bech32_from_words(words: &[bech32::u5]) -> Result<Vec<u8>, Bech32DecodingError> {
```

Comment:
This whole function is already provided by `bech32` library. Function implementation can be replaced by
```
bech32::convert_bits(words, 5, 8, true).map_err(|e| Bech32DecodingError::InvalidEncoding(e))
```

Reply:
Thank you, good catch